pkg/manager/config/proxy.go

```diff
@@ -24,46 +24,154 @@ import (
 	"go.uber.org/zap"
 )
 
+type CfgType int
+
+const (
+	CfgServer CfgType = iota
+	CfgLog
+)
+
+type OnlineCfgTypes interface {
+	config.ProxyServerOnline | config.LogOnline
+}
+
+type imeta interface {
+	getPrefix() string
+	unmarshal(bytes []byte) (any, error)
+	addToCh(any)
+	getInitial(cfg *config.Config) any
+}
+
+type meta[T OnlineCfgTypes] struct {
+	prefix   string
+	initFunc func(cfg *config.Config) T
+	ch       chan *T
+}
+
+func newMeta[T OnlineCfgTypes](prefix string, initFunc func(cfg *config.Config) T) *meta[T] {
+	return &meta[T]{
+		prefix:   prefix,
+		initFunc: initFunc,
+		ch:       make(chan *T, 1),
+	}
+}
+
+func (m meta[T]) unmarshal(bytes []byte) (any, error) {
+	var t T
+	err := json.Unmarshal(bytes, &t)
+	return &t, err
+}
+
+func (m meta[T]) getPrefix() string {
+	return m.prefix
+}
+
+func (m meta[T]) addToCh(obj any) {
+	m.ch <- obj.(*T)
+}
+
+func (m meta[T]) getInitial(cfg *config.Config) any {
+	return m.initFunc(cfg)
+}
+
+func (e *ConfigManager) initMetas() {
+	e.metas = map[CfgType]imeta{
+		CfgServer: newMeta(pathPrefixProxyServer, func(cfg *config.Config) config.ProxyServerOnline {
+			return cfg.Proxy.ProxyServerOnline
+		}),
+		CfgLog: newMeta(pathPrefixLog, func(cfg *config.Config) config.LogOnline {
+			return cfg.Log.LogOnline
+		}),
+	}
+}
```

Comment:
I appreciate the usage of generic, but this does not feel necessary to me. I mean the current implementation is not as simple as non-generics while it is not as expressive as my expectation. The core is  `map[cfgType]imeta`, which is not generic. It basically means we can archive the same effect with:

```
&imetaStruct{
  prefix: "",
  cfgGetter: func() {
  },
  unmarshaler: func() {
  },
}
```

That is much easier to grasp than generics.

Generics should have a far more expressive power. But I don't see that: we still need a lot of manual getters, setters.`GetLogConfigWatch` is not replaced by `GetConfigWatch[T]` as my expectation. Even `GetConfig[T]` still need a explicity `cfgType`.

Reply:
I considered using a non-generic struct to implement the `imeta`. However, there will be some problems:
- Each struct needs to implement a `unmarshal` function
- Each struct needs to define a channel `chan XXX`

The usage of generic types in this file is complex, but it avoids duplicated code for each configuration type, assuming that there are more and more configurations will be updated online.

Reply:
> The usage of generic types in this file is complex, but it reduces duplicated code for each configuration type, assuming that there are more and more configurations will be updated online.

That is not true. Why not just define `chan` in `ConfigManager`? I can basically archive the same effect without `imeta`:

```
struct ConfigManager {
  xxChan
  yyChan
}
func (e *ConfigManager) watchCfgProxy[T OnlineCfgTypes](ctx context.Context, cfgType string, initial *T) (<-chan *T, error)
func (e *ConfigManager) SetConfig[T OnlineCfgTypes](ctx context.Context, cfgType string, *T) error)
func (e *ConfigManager) GetConfig[T OnlineCfgTypes](ctx context.Context, cfgType string) (*T, error)
func (e *ConfigManager) GetXXChan
func (e *ConfigManager) GetYYChan
```

Considering you need three lines for each imeta definition, one line of `e.xxChan, err = e.watchCfgProxy[config.LogOnline](ctx, "log", cfg.Proxy.ProxyServerOnline)` clearly wins. There is no need to define `GetProxyConfig`, or `SetProxyConfig` too. It is used only once anyway.

Or even more aggressive:

```
struct ConfigManager {
  GetXXChan func() <-chan *XX
  ...
}
e.GetXXChan, err = e.watchCfgProxy[config.XX](ctx, "xx", cfg.XX)
```

Reply:
Do methods ( not functions) support generic types?

Reply:
> > The usage of generic types in this file is complex, but it reduces duplicated code for each configuration type, assuming that there are more and more configurations will be updated online.
> 
> That is not true. Why not just define `chan` in `ConfigManager`? I can basically archive the same effect without `imeta`:
> 
> ```
> struct ConfigManager {
>   xxChan
>   yyChan
> }
> func (e *ConfigManager) watchCfgProxy[T OnlineCfgTypes](ctx context.Context, cfgType string, initial *T) (<-chan *T, error)
> func (e *ConfigManager) SetConfig[T OnlineCfgTypes](ctx context.Context, cfgType string, *T) error)
> func (e *ConfigManager) GetConfig[T OnlineCfgTypes](ctx context.Context, cfgType string) (*T, error)
> func (e *ConfigManager) GetXXChan
> func (e *ConfigManager) GetYYChan
> ```
> 
> Considering you need three lines for each imeta definition, one line of `e.xxChan, err = e.watchCfgProxy[config.LogOnline](ctx, "log", cfg.Proxy.ProxyServerOnline)` clearly wins. There is no need to define `GetProxyConfig`, or `SetProxyConfig` too. It is used only once anyway.
> 
> Or even more aggressive:
> 
> ```
> struct ConfigManager {
>   GetXXChan func() <-chan *XX
>   ...
> }
> func (e *ConfigManager) watchCfgProxy[T OnlineCfgTypes](ctx context.Context, cfgType string, initial *T) (<-chan *T, error)
> func (e *ConfigManager) SetConfig[T OnlineCfgTypes](ctx context.Context, cfgType string, *T) error)
> func (e *ConfigManager) GetConfig[T OnlineCfgTypes](ctx context.Context, cfgType string) (*T, error)
> 
> e.GetXXChan, err = e.watchCfgProxy[config.XX](ctx, "xx", cfg.XX)
> ```

Go doesn't support type parameters on methods. It only supports type parameters on functions.

I've replaced `CfgType` to `reflect.Type` so that the interfaces don't need an extra `CfgType` parameter.